packages/typescript-estree/src/parseSettings/resolveProjectList.ts

```diff
@@ -0,0 +1,142 @@
+import debug from 'debug';
+import { sync as globSync } from 'globby';
+import isGlob from 'is-glob';
+
+import type { CanonicalPath } from '../create-program/shared';
+import {
+  createHash,
+  ensureAbsolutePath,
+  getCanonicalFileName,
+} from '../create-program/shared';
+import type { TSESTreeOptions } from '../parser-options';
+import {
+  DEFAULT_TSCONFIG_CACHE_DURATION_SECONDS,
+  ExpiringCache,
+} from './ExpiringCache';
+
+const log = debug(
+  'typescript-eslint:typescript-estree:parser:parseSettings:resovleProjectGlobs',
+);
+
+let RESOLUTION_CACHE: ExpiringCache<string, readonly CanonicalPath[]> | null =
+  null;
+
+/**
+ * Normalizes, sanitizes, resolves and filters the provided project paths
+ */
+export function resolveProjectList(
+  options: Readonly<{
+    cacheLifetime?: TSESTreeOptions['cacheLifetime'];
+    project: TSESTreeOptions['project'];
+    projectFolderIgnoreList: TSESTreeOptions['projectFolderIgnoreList'];
+    singleRun: boolean;
+    tsconfigRootDir: string;
+  }>,
+): readonly CanonicalPath[] {
+  const sanitizedProjects: string[] = [];
+
+  // Normalize and sanitize the project paths
+  if (typeof options.project === 'string') {
+    sanitizedProjects.push(options.project);
+  } else if (Array.isArray(options.project)) {
+    for (const project of options.project) {
+      if (typeof project === 'string') {
+        sanitizedProjects.push(project);
+      }
+    }
+  }
+
+  if (sanitizedProjects.length === 0) {
+    return [];
+  }
+
+  const projectFolderIgnoreList = (
+    options.projectFolderIgnoreList ?? ['**/node_modules/**']
+  )
+    .reduce<string[]>((acc, folder) => {
+      if (typeof folder === 'string') {
+        acc.push(folder);
+      }
+      return acc;
+    }, [])
+    // prefix with a ! for not match glob
+    .map(folder => (folder.startsWith('!') ? folder : `!${folder}`));
+
+  const cacheKey = getHash({
+    project: sanitizedProjects,
+    projectFolderIgnoreList,
+    tsconfigRootDir: options.tsconfigRootDir,
+  });
+  if (RESOLUTION_CACHE == null) {
+    RESOLUTION_CACHE = new ExpiringCache(
```

Comment:
Nit:

```suggestion
  if (RESOLUTION_CACHE ??= new ExpiringCache(
```

Reply:
I split it out like this because it means I can have the separate caching code path run only if the cache isn't new.

No point looking up the cache if the we know for sure the cache is empty!